File: bin/GeneExpression.pl
#!/usr/bin/perl


use strict;
use warnings;
use Getopt::Long;
use Carp;
use Interpolation 'arg:@->$' => \&argument;
use Time::HiRes qw(gettimeofday tv_interval);
use threads;
use threads::shared;

use Cwd qw(abs_path);
use FindBin;
use lib abs_path("$FindBin::Bin/../lib");


require "Perlsub.pl";

my $ANNOT = $ENV{'ANNOT'};

# Flush output after every write
select( (select(STDOUT), $| = 1 )[0] );

##
## This program
## 
## run with "--help" for usage information
##
## Robin Meyers

# Forward declarations
sub parse_command_line;
sub calculate_gene_activity;


# Global flags and arguments, 
# Set by command line arguments
my $indir;
my $outdir;
my $nonmapdir;
my $assembly = "mm9";
my $tagsize = 35;
my $max_threads = 2;
my $nodes = 8;
my $lambda = 0.1;
my $alpha = 2;
my $skipprofile;
my $upstream = 5;
my $downstream = 10;
my $binsize = 50;

# Global variables 
my $refgene;
my $chrsize;

#
# Start of Program
#

parse_command_line;

$refgene = "$ANNOT/refGene.$assembly.txt";
$chrsize = "$ANNOT/chrSize.$assembly.txt";

my %chrsize = readChromsizeFile($chrsize);

my @tagfiles = listFilesInDir($indir);
my @summaryfiles = ();
my @profiles = ();

calculate_gene_activity;





#
# End of program
#

sub calculate_gene_activity {

	my $t0 = [gettimeofday];
	my @threads = ();

	foreach my $tagfile (sort @tagfiles) {
		next unless ($tagfile =~ /\.map$/);
		
		my $chrname = substr($tagfile,-6,2) + 1;
		if ($assembly =~ /mm/) {
			next if ($chrname == 22);
			$chrname =~ s/20/X/;
			$chrname =~ s/21/Y/;
		} elsif ($assembly =~ /hg/) {
			next if ($chrname == 25);
			$chrname =~ s/23/X/;
			$chrname =~ s/24/Y/;
		}
		my $chrlen = $chrsize{"chr$chrname"};
		my $smalltagfile = "$indir/chr$chrname.txt";
		unless (-r $smalltagfile) {
			print "\nConverting $tagfile to smaller $smalltagfile\n";
			open TAG, "<", $tagfile or croak "Error: cannot open $tagfile for reading";
			open STAG, ">", $smalltagfile or croak "Error: cannot open $smalltagfile for writing";
			while (<TAG>) {
				next unless /\S/;
				chomp;
				my @tag = split("\t");
				print STAG join("\t",$tag[2],$tag[1],$tag[3]);
				print STAG "\n";
			}
		}
		
		my $summary = "$outdir/chr${chrname}-genes.txt";
		my $profile;
		push(@summaryfiles,$summary);

		unless (defined $skipprofile) {
			$profile = "$outdir/chr${chrname}-profile.txt";
			push(@profiles,$profile);
		} else {
			$profile = "NA";
		}

		my $nonmapfile = "$nonmapdir/chr$chrname-nonmappable.bed";
		

		while (1) {

			# joins any threads if possible
			foreach my $thr (@threads) {
				$thr->join() if $thr->is_joinable();
			}

			my @running = threads->list(threads::running);
			
			# if there are open threads, create a new one, push it onto list, and exit while loop
			if (scalar @running < $max_threads) {
				my $thr = threads->create( sub {
							my $t0_chr = [gettimeofday];
							print "\nChromsome $chrname\nLength: $chrlen\n";
							my $Rcmd = "$FindBin::Bin/../R/GeneExpression.R $smalltagfile $summary $refgene $nonmapfile"
								. " tagsize=$tagsize lambda=$lambda alpha=$alpha nodes=$nodes";
							$Rcmd = $Rcmd . " profile_file=$profile upstream=$upstream downstream=$downstream binsize=$binsize"
								unless (defined $skipprofile);
							System($Rcmd) or croak "Error: failed on chr$chrname during GeneExpression.R call";
							my $elapsed = tv_interval($t0_chr);
							printf("\nFinished chr%s in %.2f seconds.\n", $chrname, $elapsed);
						});
				push(@threads,$thr);
				sleep(1);
				last;
			}
			sleep(5);
		}
	
	}

	# waits for all threads to finish
	while( scalar threads->list(threads::all) > 0) {
		for my $thr (@threads) {
			$thr->join() if $thr->is_joinable;
		}
		sleep(5);
	}

	my $total = "$outdir/totalGeneExpression.txt";
	System("cat @summaryfiles | perl -p -e 'if (\$.>1){s/^Gene.*\\n//}' > $total");

	unless (defined $skipprofile) {
		my $total_profile = "$outdir/totalGeneProfile.txt";
		System("cat @profiles | perl -p -e 'if (\$.>1){s/^Gene.*\\n//}' > $total_profile");
	}

	my $elapsed = tv_interval($t0);

	printf("\nFinished calculating gene expression in %.2f seconds.\n", $elapsed);

}


sub parse_command_line {
	my $help;

	usage() if (scalar @ARGV==0);

	my $result = GetOptions ( "assembly=s" => \$assembly,
                            "threads=i" => \$max_threads,
                            "nodes=i" => \$nodes,
                            "lambda=f" => \$lambda,
                            "alpha=f" => \$alpha,
                            "tagsize=i" => \$tagsize,
                            "skipprofile" => \$skipprofile,
                            "upstream=i" => \$upstream,
                            "downstream=i" => \$downstream,
                            "binsize=i" => \$binsize,
                            "help" => \$help
				            ) ;
	
	usage() if ($help);
	usage() if (scalar @ARGV < 3);

    $indir = shift(@ARGV);
    $outdir = shift(@ARGV);
    $nonmapdir = shift(@ARGV);

    #Check options

    croak "Error: input directory $indir does not exist" unless (-d $indir);
    unless (-d $outdir) {
        croak "Error: output directory $outdir does not exist and cannot be created" unless System("mkdir $outdir");
    }
    croak "Error: nonmappable directory $nonmapdir does not exist" unless (-d $nonmapdir);

    croak "Error: unrecognizable assembly $assembly" unless $assembly =~ /^(mm8|mm9|hg18|hg19)$/;


	exit unless $result;
}


sub usage()
{
print<<EOF;
Title, by Robin Meyers, ddmonthyyyy

This program .


Usage: $0 arg1 arg2 arg3 ...
        [--option VAL] [--flag] [--help]

Arguments (defaults in parentheses):

$arg{"indir","File path containing input file for each chromosome"}
$arg{"outdir","File path to put output files"}
$arg{"nonmapdir","File path containing bedfiles of all nonmappable basepairs in genome"}
$arg{"--assembly","Genome assembly of organism",$assembly}
$arg{"--tagsize","Length of tags in bp",$tagsize}
$arg{"--lambda","Background tag density in tags/kb (on both strands)",$lambda}
$arg{"--alpha","Over-dispersion parameter from negative binomial background distribution model",$alpha}
$arg{"--threads","Number of perl threads to run on - number of chromosomes analyzed at once",$max_threads}
$arg{"--nodes","Computes activity on clusters with this many nodes in the R program - number of genes analyzed at once", $nodes}
$arg{"--skipprofile","Skip computation of high-res gene expression profile - Subsequent options are for this module"}
$arg{"--upstream","Kilobases upstream from TSS to start binning",$upstream}
$arg{"--downstream","Kilobases downstream from TSS to stop binning",$downstream}
$arg{"--binsize","Bin size in bp for high-res profile",$binsize}
$arg{"--help","This helpful help screen."}


EOF

exit 1;
}

File: R/GeneExpression.R
#!/usr/bin/Rscript

ARGS <- c(
  "tagfile", "character", "file path of bowtie ref tags",
  "output","character", "file path to write summary output to",
  "refgenefile","character","file path of refseq gene list",
  "nonmapfile","character", "file path of bed file designating unmappable regions in genome"
)

OPTS <- c(
  "tagsize","numeric",35,"Length of tags in basepairs",
  "lambda","numeric",0.1,"Mean background tags per kb (both strands) from negative binomial distribution",
  "alpha","numeric",2,"Shape/size variable from negative binomial distribution",
  "profile_file","character",NA,"File stub to write profile to",
  "upstream","numeric",5,"Kilobases upstream from TSS to begin profile",
  "downstream","numeric",10,"Kilobases downstream from TSS to begin profile",
  "binsize","numeric",50,"Bin size in bp for profile",
  "nodes","numeric",8,"Number of compute nodes to run on"
)


source_local <- function(fname){
  argv <- commandArgs(trailingOnly = FALSE)
  base_dir <- dirname(substring(argv[grep("--file=", argv)], 8))
  source(paste(base_dir, fname, sep="/"))
}

source_local("Rsub.R")
source_local("calcGeneExpr.R")
require(parallel,quietly=T,warn.conflicts=F)
require(plyr,quietly=T,warn.conflicts=F)

parseArgs("GeneExpression.R", ARGS, OPTS)
options(scipen=5)

tags <- read.delim(tagfile, sep="\t", header=FALSE, stringsAsFactors=F)

colnames(tags) <- c("Tname","Strand","Tstart")
# bowtie ref tag coordinates were 0 based; change to 1
tags$Tstart <- tags$Tstart + 1
#tags <- tags[with(tags,order(Strand,Tstart)),]


refgene <- read.delim.save(refgenefile, header=T, as.is=T, sep="\t")
colnames(refgene) <- c("Gene","Tname","Tstart","Tend","Length","RefSeq","Strand")
refgene$Tname <- paste("chr",refgene$Tname,sep="")
refgene <- refgene[refgene$Tname == tags$Tname[1] & refgene$Length >= 3000,]
refgene <- refgene[with(refgene,order(Tstart,Tend)),]

nonmap <- read.delim.save(nonmapfile, header=F, as.is=T, sep="\t")
colnames(nonmap) <- c("Tname","Tstart","Tend")

#convert lambda into tags/bp
lambda <- lambda/1000

cat("\nCalculating activity for",nrow(refgene),"genes on",tags$Tname[1],"with",nodes,"compute nodes\n")
# cl <- makeSOCKcluster(rep("localhost",nodes))
# clusterExport(cl,c("refgene","tags","nonmap","tagsize","lambda","alpha","bedCount","bedIntersect"))

ngenes <- nrow(refgene)

summary <- ldply(mclapply(1:ngenes, calcGeneExpr, mc.cores=nodes))

write.table(summary,output,sep="\t",quote=F,row.names=F)


if (profile_file == "NA") {
  profile_sense_file <- NA
}

if (! is.na(profile_file) ) {
  


  x1 <- seq(-upstream*1000,downstream*1000,binsize)
  x2 <- x1 + binsize - 1
  bins <- data.frame(x1=x1,x2=x2)

#   clusterExport(cl, c("summary","bins","binsize"))


  profile <- ldply(mclapply(1:ngenes, profileGene,mc.cores=nodes))

  write.table(profile,profile_file,sep="\t",quote=F,row.names=F)

}

# stopCluster(cl)



